pinctrl: initial support for GD32

gmarull opened the pull request:
This PR adds a new pinctrl driver for GD32 SoCs for both AF and AFIO pinmux models.

Tested on:
- GD32F450I-EVAL
- GD32F403Z-EVAL (@nandojve)

It depends on: 
- #39958
- https://github.com/zephyrproject-rtos/hal_gigadevice/pull/2

gmarull committed:
drivers: pinctrl: gd32: initial support for AF based SoCs

Add a pin control driver for GD32 SoCs using the AF model.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: gd32f450i_eval: use pinctrl

Enable usage of the pinctrl driver.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
modules: hal_gigadevice: include module headers

Include the module `include` folder to allow access to the dt-bindings
helpers.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: serial: gd32: use pinctrl

Use the pinctrl API to configure peripheral pins.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: gd32f4xx: add pinctrl and gpio nodes

Add definitions for the pinctrl and gpio nodes.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: arm: gigadevice: gd32f4xx: select AF pinmux

GD32F4XX series have AF based pinmux.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
tests: drivers: pinctrl: gd32: add DT AF parse test

Add a test to check that DT information for AF model is extracted
correctly.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
dts: arm: gigadevice: gd32f403: add afio and pinctrl

Add afio, pinctrl and gpioa...g entries for the GD32F403 series.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: pinctrl: gd32: initial support for AFIO based SoCs

Add a pin control driver for GD32 SoCs using the AFIO model.

Thanks to Gerson Fernando Budke for testing and implementation
suggestions.

Co-authored-by: Gerson Fernando Budke <gerson.budke@atl-electronics.com>
Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: gd32f403z_eval: use pinctrl

Enable usage of pinctrl.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: arm: gigadevice: gd32f403: select AFIO pinmux

GD32F403 series use AFIO pinmux model.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
tests: drivers: pinctrl: gd32: add DT AFIO parse test

Add a test to check that DT information for the AFIO model is extracted
correctly.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
@nandojve Support for AFIO (like GD32F403) should now be ready. However, I haven't been able to test it as I don't have that board (will try on VF103 using https://github.com/zephyrproject-rtos/zephyr/pull/34970)

nandojve commented:
> @nandojve Support for AFIO (like GD32F403) should now be ready. However, I haven't been able to test it as I don't have that board (will try on VF103 using #34970)

Great ! I'll try soon and report to you.

nandojve commented:
Hi @gmarull ,

I tested using `gd32f403` without sucess, no characters at serial port.
I copy `usart0_default` resulting node here to help you check if something is missing.

```
usart0_default: usart0_default {
	phandle = < 0x2 >;
	pins1 {
		pinmux = < 0x11290 >, < 0x111a0 >;
	};
};
```

nandojve commented on soc/arm/gigadevice/Kconfig:
```diff
@@ -15,6 +15,11 @@ config SOC_FAMILY_GD32_ARM
 	bool
 	select SOC_FAMILY_GD32
 
+config GD32_HAS_AF_PINMUX
+	bool
+	help
+	  This option should be selected if the series use an AF pinmux model.
+
```
I think it is better move to modules/hal_gigadevice. That way we can share between ARM and RISC-V

gmarull replied:
done

nandojve commented on soc/arm/gigadevice/Kconfig:
```diff
@@ -20,6 +20,11 @@ config GD32_HAS_AF_PINMUX
 	help
 	  This option should be selected if the series use an AF pinmux model.
 
+config GD32_HAS_AFIO_PINMUX
+	bool
+	help
+	  This option should be selected if the series use an AFIO pinmux model.
+
```
Same as `GD32_HAS_AF_PINMUX`, better move to modules/haL_gigadevice

gmarull replied:
done, makes sense to cover RISC-V

nandojve commented on boards/arm/gd32f450i_eval/gd32f450i_eval_defconfig:
```diff
@@ -9,6 +9,8 @@ CONFIG_ARM_MPU=y
 CONFIG_HW_STACK_PROTECTION=y
 CONFIG_CORTEX_M_SYSTICK=y
 
+CONFIG_PINCTRL=y
+
```
As far I understand there is no practical application without use at least 1 gpio pin. This means that PINCTRL must be always enabled. My suggestion is enable at HAS_GD32_HAL.

```diff
config HAS_GD32_HAL
	bool
	select HAS_CMSIS_CORE if SOC_FAMILY_GD32_ARM
+	select PINCTRL
```

gmarull replied:
I have added 
```
config PINCTRL
	default y
```
soc/arm/gigadevice/Kconfig.defconfig, it's the usual way to do this. The hal and pinctrl are not related, other than pinctrl drivers use the hal.

nandojve commented on drivers/serial/Kconfig.gd32:
```diff
@@ -7,7 +7,7 @@ DT_COMPAT_GIGADEVICE_GD32_USART := gd,gd32-usart
 	bool "GD32 serial driver"
 	default $(dt_compat_enabled,$(DT_COMPAT_GIGADEVICE_GD32_USART))
-	depends on SOC_FAMILY_GD32
+	depends on SOC_FAMILY_GD32 && PINCTRL
```
The SOC_FAMILY_GD32 config will always enable PINCTRL.

gmarull replied:
done

nandojve commented on boards/arm/gd32f403z_eval/gd32f403z_eval_defconfig:
```diff
@@ -9,6 +9,8 @@ CONFIG_ARM_MPU=y
 CONFIG_HW_STACK_PROTECTION=y
 CONFIG_CORTEX_M_SYSTICK=y
 
+CONFIG_PINCTRL=y
+
```
You can drop here.

gmarull replied:
done

nandojve reviewed (changes requested):
Hi @gmarull ,

Excellent work!

I start to look from manifest update and have only small comments.

nandojve commented on dts/arm/gigadevice/gd32f403/gd32f403.dtsi:
```diff
@@ -46,6 +46,79 @@
+			gpioa: gpio@40010800 {
+				compatible = "gd,gd32-gpio";
+				reg = <0x40010800 0x400>;
+				rcu-periph-clock = <0xc00>;
```
```diff
git diff dts/arm/gigadevice/gd32f403/gd32f403.dtsi
diff --git a/dts/arm/gigadevice/gd32f403/gd32f403.dtsi b/dts/arm/gigadevice/gd32f403/gd32f403.dtsi
index 6fec4a1e55..9aa5ea2e09 100644
--- a/dts/arm/gigadevice/gd32f403/gd32f403.dtsi
+++ b/dts/arm/gigadevice/gd32f403/gd32f403.dtsi
@@ -65,7 +65,7 @@
                        gpioa: gpio@40010800 {
                                compatible = "gd,gd32-gpio";
                                reg = <0x40010800 0x400>;
-                               rcu-periph-clock = <0xc00>;
+                               rcu-periph-clock = <0x602>;
                                status = "disabled";
                                label = "GPIOA";
                        };
@@ -73,7 +73,7 @@
                        gpiob: gpio@40010c00 {
                                compatible = "gd,gd32-gpio";
                                reg = <0x40010c00 0x400>;
-                               rcu-periph-clock = <0xc01>;
+                               rcu-periph-clock = <0x603>;
                                status = "disabled";
                                label = "GPIOB";
                        };
@@ -81,7 +81,7 @@
                        gpioc: gpio@40011000 {
                                compatible = "gd,gd32-gpio";
                                reg = <0x40011000 0x400>;
-                               rcu-periph-clock = <0xc02>;
+                               rcu-periph-clock = <0x604>;
                                status = "disabled";
                                label = "GPIOC";
                        };
@@ -89,7 +89,7 @@
                        gpiod: gpio@40011400 {
                                compatible = "gd,gd32-gpio";
                                reg = <0x40011400 0x400>;
-                               rcu-periph-clock = <0xc03>;
+                               rcu-periph-clock = <0x605>;
                                status = "disabled";
                                label = "GPIOD";
                        };
@@ -97,7 +97,7 @@
                        gpioe: gpio@40011800 {
                                compatible = "gd,gd32-gpio";
                                reg = <0x40011800 0x400>;
-                               rcu-periph-clock = <0xc04>;
+                               rcu-periph-clock = <0x606>;
                                status = "disabled";
                                label = "GPIOE";
                        };
@@ -105,7 +105,7 @@
                        gpiof: gpio@40011c00 {
                                compatible = "gd,gd32-gpio";
                                reg = <0x40011c00 0x400>;
-                               rcu-periph-clock = <0xc05>;
+                               rcu-periph-clock = <0x607>;
                                status = "disabled";
                                label = "GPIOF";
                        };
@@ -113,7 +113,7 @@
                        gpiog: gpio@40012000 {
                                compatible = "gd,gd32-gpio";
                                reg = <0x40012000 0x400>;
-                               rcu-periph-clock = <0xc06>;
+                               rcu-periph-clock = <0x608>;
                                status = "disabled";
                                label = "GPIOG";
                        };

```

gmarull replied:
adjusted

nandojve commented on drivers/pinctrl/pinctrl_gd32_afio.c:
```diff
@@ -0,0 +1,210 @@
+		reg &= ~GPIO_MODE_MASK(pin_num);
+
+		/* configure output speed */
+		if (mode == GD32_MODE_ALTERNATE) {
+			uint8_t speed = GD32_OSPEED_GET(pin);
+
+			speed = configure_spd(port, pin_num, speed);
+			reg |= GPIO_MODE_SET(pin_num, speed);
+		}
+
+		GPIO_CTL0(port) = reg;
+	} else {
+		uint32_t reg = GPIO_CTL1(port);
+
+		reg &= ~GPIO_MODE_MASK(pin_num - 8U);
+
+		/* configure output speed */
+		if (mode == GD32_MODE_ALTERNATE) {
+			uint8_t speed = GD32_OSPEED_GET(pin);
+
+			speed = configure_spd(port, pin_num, speed);
+			reg |= GPIO_MODE_SET(pin_num - 8U, speed);
+		}
+
+		GPIO_CTL1(port) = reg;
+	}
+
+	/* configure pull-up/down if input */
+	if (mode == GD32_MODE_GPIO_IN) {
+		uint8_t pupd = GD32_PUPD_GET(pin);
+
+		if (pupd == GD32_PUPD_PULLDOWN) {
+			GPIO_BC(port) = BIT(pin_num);
+		} else if (pupd == GD32_PUPD_PULLUP) {
+			GPIO_BOP(port) = BIT(pin_num);
+		}
+	}
+
+	rcu_periph_clock_disable(rcu);
+}
```
Hi @gmarull ,

The code as is not select Output AFIO mode at CTLx[1:0], where x is the pin_num. I made some changes based in the below table. After that, USART start to work. I noted that input-floating (bias-high-impedance) is missing.
My feeling is that `configure_pin` function should be reviewed to match with GPIO configuration table.

![GPIO_configuration_table](https://user-images.githubusercontent.com/7027887/141598514-a293aec5-25de-43eb-b150-22979bcc42b5.jpg)

```diff
git diff drivers/pinctrl/pinctrl_gd32_afio.c
diff --git a/drivers/pinctrl/pinctrl_gd32_afio.c b/drivers/pinctrl/pinctrl_gd32_afio.c
index 613e065d8c..a3ad522e6c 100644
--- a/drivers/pinctrl/pinctrl_gd32_afio.c
+++ b/drivers/pinctrl/pinctrl_gd32_afio.c
@@ -102,7 +102,8 @@ static inline uint8_t configure_spd(uint32_t port, uint8_t pin_num,
 static void configure_pin(pinctrl_soc_pin_t pin)
 {
        uint8_t port_idx, mode, pin_num;
-       uint32_t rcu, port;
+       uint32_t rcu, port, reg_tmp;
+       volatile uint32_t *reg;
 
        port_idx = GD32_PORT_GET(pin);
        __ASSERT_NO_MSG(port_idx < ARRAY_SIZE(gd32_port_addrs));
@@ -111,50 +112,35 @@ static void configure_pin(pinctrl_soc_pin_t pin)
        port = gd32_port_addrs[port_idx];
        pin_num = GD32_PIN_GET(pin);
        mode = GD32_MODE_GET(pin);
+       reg = (pin_num < 8U) ? &GPIO_CTL0(port) : &GPIO_CTL1(port);
+       pin_num = (pin_num < 8U) ? pin_num : (pin_num - 8U);
 
        rcu_periph_clock_enable(rcu);
 
-       if (pin_num < 8U) {
-               uint32_t reg = GPIO_CTL0(port);
+       reg_tmp = *reg;
+       reg_tmp &= ~GPIO_MODE_MASK(pin_num);
 
-               reg &= ~GPIO_MODE_MASK(pin_num);
+       if (mode == GD32_MODE_ALTERNATE) {
+               uint8_t speed = GD32_OSPEED_GET(pin);
 
-               /* configure output speed */
-               if (mode == GD32_MODE_ALTERNATE) {
-                       uint8_t speed = GD32_OSPEED_GET(pin);
+               speed = configure_spd(port, pin_num, speed);
+               reg_tmp |= GPIO_MODE_SET(pin_num, speed | 0x08);
 
-                       speed = configure_spd(port, pin_num, speed);
-                       reg |= GPIO_MODE_SET(pin_num, speed);
-               }
-
-               GPIO_CTL0(port) = reg;
-       } else {
-               uint32_t reg = GPIO_CTL1(port);
-
-               reg &= ~GPIO_MODE_MASK(pin_num - 8U);
-
-               /* configure output speed */
-               if (mode == GD32_MODE_ALTERNATE) {
-                       uint8_t speed = GD32_OSPEED_GET(pin);
-
-                       speed = configure_spd(port, pin_num, speed);
-                       reg |= GPIO_MODE_SET(pin_num - 8U, speed);
-               }
-
-               GPIO_CTL1(port) = reg;
-       }
-
-       /* configure pull-up/down if input */
-       if (mode == GD32_MODE_GPIO_IN) {
+               /* TODO: OPEN-DRAIN */
+       } else if (mode == GD32_MODE_GPIO_IN) {
                uint8_t pupd = GD32_PUPD_GET(pin);
 
                if (pupd == GD32_PUPD_PULLDOWN) {
                        GPIO_BC(port) = BIT(pin_num);
+                       reg_tmp |= GPIO_MODE_SET(pin_num, 0x08); /* PUPD Mode */
                } else if (pupd == GD32_PUPD_PULLUP) {
                        GPIO_BOP(port) = BIT(pin_num);
+                       reg_tmp |= GPIO_MODE_SET(pin_num, 0x08); /* PUPD Mode */
                }
        }
 
+       *reg = reg_tmp;
+
        rcu_periph_clock_disable(rcu);
 }
```

gmarull replied:
Thanks for the testing and suggestions, seems like I missed quite a few things. I have updated the driver, I hope all things are covered now.

nandojve commented on dts/bindings/pinctrl/gd,gd32-pinctrl-common.yaml:
```diff
@@ -0,0 +1,26 @@
+      child-binding:
+        child-binding:
+          property-allowlist:
+            - drive-push-pull
+            - drive-open-drain
+            - bias-disable
+            - bias-pull-down
+            - bias-pull-up
```
bias-high-impedance is missing

gmarull replied:
This option is not required, same as STM32 (see https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/pinctrl/st%2Cstm32-pinctrl.yaml#L18). Not setting any bias and operating as input would be sufficient as far as I understand.

nandojve replied:
> Not setting any bias and operating as input would be sufficient as far as I understand.

I understand not setting any bias, which means `bias-disable` (default state), equals `Analog`. We should not work with a hidden state because people can made confusion. The configuration table have 4 options for Input and there are 3 options defined. If you look at https://github.com/zephyrproject-rtos/zephyr/blob/a5c2a6169b65849b48e3416c16b786aa53078562/dts/bindings/pinctrl/pincfg-node.yaml#L25-L28 this is the option that Linux define for the `Input Floating`.

gmarull replied:
The problem is that analog is treated via the pinmux value as a separate mode, not as an input mode. In case of DAC (an output signal), ANALOG still has to be used. For low-power mode, what is usually done is to set the analog mode (default config) on a pin, this is why I added the analog option for all pins in the generator:

```
/* ANALOG */
#define ANALOG_PA0 \
	GD32_PINMUX_AFIO('A', 0, ANALOG, NORMP)
#define ANALOG_PA1 \
	GD32_PINMUX_AFIO('A', 1, ANALOG, NORMP)
...
```

I know it can be a bit confusing, it's really STM32 inheritance (it's done this way, see for example https://elixir.bootlin.com/linux/latest/source/arch/arm/boot/dts/stm32mp15-pinctrl.dtsi#L62). I think that treating analog in a special way makes things more understandable when it comes to the analog signals, but not that clear for low-power modes. In such case, it would probably make more sense to configure the pin as an input and set bias-high-impedance.

nandojve replied:
In this case, it is important give some background for users. The information you presented at last comment is relevant to users understand how this should be configured. Other important thing is that no one can determine from where user will start to read the documentation. I think it could be very valuable improve documentation and try connect those things. For instance, correlate the _GPIO configuration table_ with the options available can be very valuable to users. The same way, explain to then the why Analog is configured differently will help. Maybe provide pointer that exists in Zephyr can help to. Take in consideration that a lot of users don't know Linux and devicetree.

gmarull replied:
I have updated the bindings description explaining more details on low power modes, please check.

nandojve commented on dts/bindings/pinctrl/gd,gd32-pinctrl-af.yaml:
```diff
@@ -0,0 +1,71 @@
+
+    The list of supported standard properties:
+
+    - drive-push-pull: Push-pull drive mode (default, not required).
+    - drive-open-drain: Open-drain drive mode.
+    - bias-disable: Disable pull-up/down (default, not required).
+    - bias-pull-up: Enable pull-up resistor.
+    - bias-pull-down: Enable pull-down resistor.
+    - slew-rate: Set the maximum speed (and so the slew-rate) of the output
+      signal (default: 2MHz).
```
bias-high-impedance is missing

gmarull replied:
ditto

nandojve replied:
> bias-disable: Disable pull-up/down (default, not required).

What is this? Analog or High-Impedance? Again, it is not clear the meaning and it is possible improve here to a very clear pin state.

nandojve commented on dts/bindings/pinctrl/gd,gd32-pinctrl-afio.yaml:
```diff
@@ -0,0 +1,98 @@
+
+    The list of supported standard properties:
+
+    - drive-push-pull: Push-pull drive mode (default, not required). Only
+      applies for GPIO_IN mode.
+    - drive-open-drain: Open-drain drive mode. Only applies for GPIO_IN mode.
+    - bias-disable: Disable pull-up/down (default, not required). Only applies
+      for GPIO_IN mode.
+    - bias-pull-up: Enable pull-up resistor. Only applies for GPIO_IN mode.
+    - bias-pull-down: Enable pull-down resistor. Only applies for GPIO_IN mode.
+    - slew-rate: Set the maximum speed (and so the slew-rate) of the output
+      signal (default: 2MHz). Only applies for ALTERNATE mode.
```
bias-high-impedance is missing

gmarull replied:
ditto

nandojve reviewed (changes requested):
Hi @gmarull ,

After some fixes AFIO is working:

```
*** Booting Zephyr OS build v2.7.99-1382-g31b26bab517c  ***
Running test suite pinctrl_gd32
===================================================================
START - test_dt_extract
 PASS - test_dt_extract in 0.1 seconds
===================================================================
Test suite pinctrl_gd32 succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

gmarull committed:
soc: arm: gigadevice: enable pinctrl by default

Enable pinctrl by default, since it is an essential component on almost
every firmware. Inclusion of series defconfig has also been guarded with
SoC availability (was missing).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
@cameled reviews welcome

nandojve commented:
The AFIO is working now!

nandojve reviewed (approved):
Hi @gmarull ,

Thank you for this PR!

mbolivar-nordic reviewed (approved):
I didn't review this very closely, but it looks like it's following all the latest practices from a quick skim.
